From: Tim Waugh <twaugh@redhat.com>
Date: Mon, 8 Nov 2004 09:33:22 -0500
Subject: Automatically disable DFA when processing multibyte input

grep is painfully slow on multibyte locales. Slowdown factor >30
observed.  This patch makes the use of the DFA conditional on
whether the current locale character encoding is a multibyte one.
For UTF-8, the DFA is turned off.  This way, the required
multibyte processing (mbrtowc etc) reduces to almost none; it is
just not generally necessary as far as grep is concerned.

GREP_USE_DFA environment variable overrides.

The longer-term solution is to make grep use the system regex for
multibyte encodings.  The GNU libc implementation has quite an
efficient implementation now.

This patch requires 64-egf-speedup.patch

Bug-Fedora: https://bugzilla.redhat.com/show_bug.cgi?id=121313
Benchmarks: http://thread.gmane.org/gmane.comp.gnu.grep.bugs/739/focus=744
Benchmarks: http://thread.gmane.org/gmane.comp.gnu.grep.bugs/100
Bug-Debian: http://bugs.debian.org/181378
---
 src/search.c |   31 ++++++++++++++++++++++++++++---
 1 files changed, 28 insertions(+), 3 deletions(-)

diff --git a/src/search.c b/src/search.c
index b493452..f6ff4f7 100644
--- a/src/search.c
+++ b/src/search.c
@@ -327,6 +327,8 @@ EXECUTE_FCT(EGexecute)
   char eol = eolbyte;
   int backref, start, len, best_len;
   struct kwsmatch kwsm;
+  static int use_dfa;
+  static int use_dfa_checked = 0;
   size_t i, ret_val;
 #ifdef MBS_SUPPORT
   int mb_cur_max = MB_CUR_MAX;
@@ -334,6 +336,26 @@ EXECUTE_FCT(EGexecute)
   memset (&mbs, '\0', sizeof (mbstate_t));
 #endif /* MBS_SUPPORT */
 
+  if (!use_dfa_checked)
+    {
+      char *grep_use_dfa = getenv ("GREP_USE_DFA");
+      if (!grep_use_dfa)
+	{
+#ifdef MBS_SUPPORT
+	  /* Turn off DFA when processing multibyte input. */
+	  use_dfa = (MB_CUR_MAX == 1);
+#else
+	  use_dfa = 1;
+#endif /* MBS_SUPPORT */
+	}
+      else
+	{
+	  use_dfa = atoi (grep_use_dfa);
+	}
+
+      use_dfa_checked = 1;
+    }
+
   buflim = buf + size;
 
   for (beg = end = buf; end < buflim; beg = end)
@@ -402,7 +424,8 @@ EXECUTE_FCT(EGexecute)
 #endif /* MBS_SUPPORT */
 		  (kwsm.index < kwset_exact_matches))
 		goto success;
-	      if (dfaexec (&dfa, beg, end - beg, &backref) == (size_t) -1)
+	      if (use_dfa &&
+		  dfaexec (&dfa, beg, end - beg, &backref) == (size_t) -1)
 		continue;
 	    }
 	  else
@@ -411,7 +434,9 @@ EXECUTE_FCT(EGexecute)
 #ifdef MBS_SUPPORT
 	      size_t bytes_left = 0;
 #endif /* MBS_SUPPORT */
-	      size_t offset = dfaexec (&dfa, beg, buflim - beg, &backref);
+	      size_t offset = 0;
+	      if (use_dfa)
+		offset = dfaexec (&dfa, beg, buflim - beg, &backref);
 	      if (offset == (size_t) -1)
 		break;
 	      /* Narrow down to the line we've found. */
@@ -453,7 +478,7 @@ EXECUTE_FCT(EGexecute)
 		--beg;
 	    }
 	  /* Successful, no backreferences encountered! */
-	  if (!backref)
+	  if (use_dfa && !backref)
 	    goto success;
 	}
       else
-- 
1.6.6

